File: src/mission/controlMissionLambda.ts
import { SQSEvent, SQSHandler, SQSRecord } from 'aws-lambda';
import { DynamoDBClient, PutItemCommand, GetItemCommand, GetItemCommandInput, PutItemCommandInput } from '@aws-sdk/client-dynamodb';

const dynamoClient = new DynamoDBClient({
    region: 'eu-central-1'
})

const handler: SQSHandler = async (event: SQSEvent): Promise<void> => {
    for (const message of event.Records) {
        await processMessage(message)
    }
}

async function processMessage(message: SQSRecord) {
    const launch: { rocketName: string } = JSON.parse(message.body)
    if (!launch.rocketName) {
        throw Error('No rocket name specified')
    }

    const input: GetItemCommandInput = {
        TableName: 'Mission',
        Key: {
            'RocketName': {
                'S': launch.rocketName
            },
        },
    }
    const response = await dynamoClient.send(new GetItemCommand(input))
    if (!response.Item) {
        const createInput: PutItemCommandInput = {
            TableName: 'Mission',
            Item: {
                'RocketName': {
                    'S': launch.rocketName
                },
                'Destination': {
                    'S': 'Mars'
                },
                'Progress': {
                    'N': '0'
                }
            }
        }
        await dynamoClient.send(new PutItemCommand(createInput))

    }
}

module.exports = { handler }

File: src/mission/control_mission_lambda.py
import json

import boto3


def handler(event, context):
    rocket_name = event["pathParameters"]["rocketName"]
    if not rocket_name:
        raise ValueError("No rocket name specified")

    client = boto3.client("dynamodb", region_name="eu-central-1")

    response = client.get_item(
        TableName="Mission", Key={"RocketName": {"S": rocket_name}}
    )

    if "Item" not in response:
        client.put_item(
            TableName="Mission",
            Item={
                "RocketName": {"S": rocket_name},
                "Destination": {"S": "Mars"},
                "Progress": {"N": "0"},
            },
        )

    return {
        "statusCode": 200,
        "body": json.dumps({"destination": "Mars", "progress": 10}),
    }
